debezium-server/debezium-server-pravega/src/main/java/io/debezium/server/pravega/PravegaTxnChangeConsumer.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.server.pravega;
+
+import java.net.URI;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.Dependent;
+import javax.inject.Named;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.engine.ChangeEvent;
+import io.debezium.engine.DebeziumEngine.ChangeConsumer;
+import io.debezium.engine.DebeziumEngine.RecordCommitter;
+import io.debezium.server.BaseChangeConsumer;
+import io.pravega.client.ClientConfig;
+import io.pravega.client.EventStreamClientFactory;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.Transaction;
+import io.pravega.client.stream.TransactionalEventStreamWriter;
+import io.pravega.client.stream.TxnFailedException;
+import io.pravega.client.stream.impl.ByteArraySerializer;
+
+@Named("pravega-txn")
```

Comment:
LGTM, still I'd prefer to have only one sink, `pravega` and there would be a config option that would decide whether the tx will be used or not. I don't think it is necessary to provide two sinks to handle those two use cases.

Reply:
@jpechane -- I can bring these two sinks into one sink.

I like how two sinks separated the used/unused member variables by use-case, but I can achieve the same result with inner classes or similar.

Perhaps I'll use a boolean config property to turn transactions on or off.

I would like the transactional sink to achieve exactly-once semantics in the short-term, but I think I need some support from Debezium to get there. The transactional approach does minimize the window where failure can produce duplicate messages, but the two approaches will have different performance profiles, which I'll look to characterize eventually.

Reply:
This will unfortunately still take some time https://issues.apache.org/jira/browse/KAFKA-6080

Reply:
@jpechane -- Does Debezium Server also rely on completion of "Transactional EoS for source connectors." Doesn't Debezium Server effectively mock Kafka Connect? I was assuming that gives Debezium Server more freedom of action, but maybe the duplicates still come from the Connect source even if we can squash them coming from the interactions of the Debezium batches and the Pravega sink.

In the meantime, I think we can make Debezium Server exactly-once with respect to Debezium batches and Debezium Server sinks. (Accepting that duplicates could still come in across batches via recovery-handling on the source-side.)

Reply:
@derekm ok, I think this discussion should be moved elswhere, I propose to create a follow up Jira for improvements.
But generally I understand now what do you want to achieve. There is a missing piece in the puzzle.
You need to implement an instance of `org.apache.kafka.connect.storage.OffsetBackingStore` that would store the connector offsets in Pravega, If you'd write data batch *and* ofssets to Pravega in a single transaction then you will offer exactly-once delivery.